src/Sulu/Bundle/MediaBundle/Resources/js/stores/MediaUploadStore/MediaUploadStore.js

```diff
@@ -0,0 +1,84 @@
+// @flow
+import {action, computed, observable} from 'mobx';
+import {ResourceStore, ResourceMetadataStore} from 'sulu-admin-bundle/stores';
+import {ResourceRequester} from 'sulu-admin-bundle/services';
+
+const RESOURCE_KEY = 'media';
+const THUMBNAIL_SIZE = 'sulu-400x400';
+
+export default class MediaUploadStore {
+    resourceStore: ResourceStore;
+    @observable uploading: boolean;
+    @observable progress: number;
+
+    constructor(resourceStore: ResourceStore) {
+        this.resourceStore = resourceStore;
+    }
+
+    @computed get source(): ?string {
+        if (!this.resourceStore.data.id) {
+            return;
+        }
+
+        const {
+            data: {
+                thumbnails: {
+                    [THUMBNAIL_SIZE]: thumbnailPath,
+                },
+            },
+        } = this.resourceStore;
+
+        return `${window.location.origin}${thumbnailPath}`;
+    }
+
+    @action setUploading(uploading: boolean) {
+        this.uploading = uploading;
+
+        if (!uploading) {
+            // Using timeout because mobx async behavior is messing up animations
+            // which are based on the upload progress
+            setTimeout(() => this.setProgress(0), 1000);
+        }
+    }
+
+    @action setProgress(progress: number) {
+        this.progress = Math.ceil(progress);
+    }
+
+    update(mediaId: string | number, file: File) {
+        const baseUrl = ResourceMetadataStore.getBaseUrl(RESOURCE_KEY);
+        const queryString = ResourceRequester.buildQueryString({
+            action: 'new-version',
+            locale: this.resourceStore.locale.get(),
+        });
+        const url = baseUrl + '/' + mediaId + queryString;
+
+        this.setUploading(true);
+        this.upload(file, url)
+            .then((data: Object) => {
+                for (const key of Object.keys(data)) {
+                    this.resourceStore.set(key, data[key]);
+                }
+                this.setUploading(false);
+            });
+    }
+
+    upload(file: File, url: string): Promise<*> {
+        return new Promise((resolve, reject) => {
+            const xhr = new XMLHttpRequest();
+            const form = new FormData();
+
+            xhr.open('POST', url);
```

Comment:
Any reason not to use `fetch` here?

Reply:
fetch doesn't provide a `progress` event